backend/src/schema/resolvers/users.spec.js

```diff
@@ -1,276 +1,465 @@
-        const expected = '"name" length must be at least 3 characters long'
-        await expect(client.request(mutation, variables)).rejects.toThrow(expected)
-      })
+    it('with too short name', async () => {
+      const variables = {
+        id: 'u47',
+        name: '  ',
+      }
+      const { errors } = await mutate({ mutation: updateUserMutation, variables })
+      expect(errors[0]).toHaveProperty(
+        'message',
+        'child "name" fails because ["name" length must be at least 3 characters long]',
+      )
     })
   })
+})
 
-  describe('DeleteUser', () => {
-    let deleteUserVariables
-    let asAuthor
-    const deleteUserMutation = gql`
-      mutation($id: ID!, $resource: [Deletable]) {
-        DeleteUser(id: $id, resource: $resource) {
+describe('DeleteUser', () => {
+  const deleteUserMutation = gql`
+    mutation($id: ID!, $resource: [Deletable]) {
+      DeleteUser(id: $id, resource: $resource) {
+        id
+        name
+        about
+        deleted
+        contributions {
           id
-          contributions {
-            id
-            deleted
-          }
+          content
+          contentExcerpt
+          deleted
           comments {
             id
+            content
+            contentExcerpt
             deleted
           }
         }
+        comments {
+          id
+          content
+          contentExcerpt
+          deleted
+        }
       }
-    `
+    }
+  `
+  beforeEach(async () => {
+    variables = { id: ' u343', resource: [] }
+
+    user = await factory.create('User', {
+      name: 'My name should be deleted',
+      about: 'along with my about',
+      id: 'u343',
+    })
+    await factory.create('User', {
+      email: 'friends-account@example.org',
+      password: '1234',
+      id: 'not-my-account',
+    })
+  })
+
+  describe('unauthenticated', () => {
+    it('throws authorization error', async () => {
+      const { errors } = await mutate({ mutation: deleteUserMutation, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
+    })
+  })
+
+  describe('authenticated', () => {
     beforeEach(async () => {
-      await factory.create('User', {
-        email: 'test@example.org',
-        password: '1234',
-        id: 'u343',
-      })
-      await factory.create('User', {
-        email: 'friends-account@example.org',
-        password: '1234',
-        id: 'u565',
-      })
-      deleteUserVariables = { id: 'u343', resource: [] }
+      authenticatedUser = await user.toJson()
     })
 
-    describe('unauthenticated', () => {
-      it('throws authorization error', async () => {
-        client = new GraphQLClient(host)
-        await expect(client.request(deleteUserMutation, deleteUserVariables)).rejects.toThrow(
-          'Not Authorised',
-        )
+    describe("attempting to delete another user's account", () => {
+      beforeEach(() => {
+        variables = { ...variables, id: 'not-my-account' }
       })
-    })
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-03T11:44:09Z" title="Tuesday, September 3rd 2019, 1:44:09 pm +02:00">Sep 3, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,273 +1,441 @@
-import { GraphQLClient } from 'graphql-request'
 import Factory from '../../seed/factories'
-import { host, login, gql } from '../../jest/helpers'
-import { neode } from '../../bootstrap/neo4j'
+import { gql } from '../../jest/helpers'
+import { neode as getNeode, getDriver } from '../../bootstrap/neo4j'
+import createServer from '../../server'
+import { createTestClient } from 'apollo-server-testing'
 
-let client
 const factory = Factory()
-const instance = neode()
 const categoryIds = ['cat9']
+let user
 
-afterEach(async () => {
-  await factory.cleanDatabase()
-})
-
-describe('users', () => {
-  describe('User', () => {
-    describe('query by email address', () => {
-      beforeEach(async () => {
-        await factory.create('User', { name: 'Johnny', email: 'any-email-address@example.org' })
-      })
+let query
+let mutate
+let authenticatedUser
+let variables
 
-      const query = `query($email: String) { User(email: $email) { name } }`
-      const variables = { email: 'any-email-address@example.org' }
-      beforeEach(() => {
-        client = new GraphQLClient(host)
-      })
+const driver = getDriver()
+const neode = getNeode()
 
-      it('is forbidden', async () => {
-        await expect(client.request(query, variables)).rejects.toThrow('Not Authorised')
-      })
+beforeAll(() => {
+  const { server } = createServer({
+    context: () => {
+      return {
+        driver,
+        neode,
+        user: authenticatedUser,
+      }
+    },
+  })
+  query = createTestClient(server).query
+  mutate = createTestClient(server).mutate
+})
 
-      describe('as admin', () => {
-        beforeEach(async () => {
-          const userParams = {
-            role: 'admin',
-            email: 'admin@example.org',
-            password: '1234',
-          }
-          const factory = Factory()
-          await factory.create('User', userParams)
-          const headers = await login(userParams)
-          client = new GraphQLClient(host, { headers })
-        })
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
 
-        it('is permitted', async () => {
-          await expect(client.request(query, variables)).resolves.toEqual({
-            User: [{ name: 'Johnny' }],
-          })
-        })
-      })
+describe('User', () => {
+  describe('query by email address', () => {
+    beforeEach(async () => {
+      await factory.create('User', { name: 'Johnny', email: 'any-email-address@example.org' })
     })
-  })
 
-  describe('UpdateUser', () => {
-    const userParams = {
-      email: 'user@example.org',
-      password: '1234',
-      id: 'u47',
-      name: 'John Doe',
-    }
-    const variables = {
-      id: 'u47',
-      name: 'John Doughnut',
-    }
-
-    const mutation = `
-      mutation($id: ID!, $name: String) {
-        UpdateUser(id: $id, name: $name) {
-          id
+    const userQuery = gql`
+      query($email: String) {
+        User(email: $email) {
           name
         }
       }
     `
+    const variables = { email: 'any-email-address@example.org' }
 
-    beforeEach(async () => {
-      await factory.create('User', userParams)
+    it('is forbidden', async () => {
+      const { errors } = await query({ query: userQuery, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
     })
 
-    describe('as another user', () => {
+    describe('as admin', () => {
       beforeEach(async () => {
-        const someoneElseParams = {
-          email: 'someone-else@example.org',
+        const userParams = {
+          role: 'admin',
+          email: 'admin@example.org',
           password: '1234',
-          name: 'James Doe',
         }
-
-        await factory.create('User', someoneElseParams)
-        const headers = await login(someoneElseParams)
-        client = new GraphQLClient(host, { headers })
+        const admin = await factory.create('User', userParams)
+        authenticatedUser = await admin.toJson()
       })
 
-      it('is not allowed to change other user accounts', async () => {
-        await expect(client.request(mutation, variables)).rejects.toThrow('Not Authorised')
+      it('is permitted', async () => {
+        await expect(query({ query: userQuery, variables })).resolves.toMatchObject({
+          data: { User: [{ name: 'Johnny' }] },
+        })
       })
     })
+  })
+})
 
-    describe('as the same user', () => {
-      beforeEach(async () => {
-        const headers = await login(userParams)
-        client = new GraphQLClient(host, { headers })
-      })
+describe('UpdateUser', () => {
+  const userParams = {
+    email: 'user@example.org',
+    password: '1234',
+    id: 'u47',
+    name: 'John Doe',
+  }
+  const variables = {
+    id: 'u47',
+    name: 'John Doughnut',
+  }
+
+  const updateUserMutation = gql`
+    mutation($id: ID!, $name: String) {
+      UpdateUser(id: $id, name: $name) {
+        id
+        name
+      }
+    }
+  `
+
+  beforeEach(async () => {
+    user = await factory.create('User', userParams)
+  })
+
+  describe('as another user', () => {
+    beforeEach(async () => {
+      const someoneElseParams = {
+        email: 'someone-else@example.org',
+        password: '1234',
+        name: 'James Doe',
+      }
+
+      const someoneElse = await factory.create('User', someoneElseParams)
+      authenticatedUser = await someoneElse.toJson()
+    })
 
-      it('name within specifications', async () => {
-        const expected = {
+    it('is not allowed to change other user accounts', async () => {
+      const { errors } = await mutate({ mutation: updateUserMutation, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
+    })
+  })
+
+  describe('as the same user', () => {
+    beforeEach(async () => {
+      authenticatedUser = await user.toJson()
+    })
+
+    it('name within specifications', async () => {
+      const expected = {
+        data: {
           UpdateUser: {
             id: 'u47',
             name: 'John Doughnut',
           },
-        }
-        await expect(client.request(mutation, variables)).resolves.toEqual(expected)
-      })
+        },
+      }
+      await expect(mutate({ mutation: updateUserMutation, variables })).resolves.toMatchObject(
+        expected,
+      )
+    })
 
-      it('with `null` as name', async () => {
-        const variables = {
-          id: 'u47',
-          name: null,
-        }
-        const expected = '"name" must be a string'
-        await expect(client.request(mutation, variables)).rejects.toThrow(expected)
-      })
+    it('with `null` as name', async () => {
+      const variables = {
+        id: 'u47',
+        name: null,
+      }
+      const { errors } = await mutate({ mutation: updateUserMutation, variables })
+      expect(errors[0]).toHaveProperty(
+        'message',
+        'child "name" fails because ["name" contains an invalid value, "name" must be a string]',
+      )
+    })
 
-      it('with too short name', async () => {
-        const variables = {
-          id: 'u47',
-          name: '  ',
-        }
-        const expected = '"name" length must be at least 3 characters long'
-        await expect(client.request(mutation, variables)).rejects.toThrow(expected)
-      })
+    it('with too short name', async () => {
+      const variables = {
+        id: 'u47',
+        name: '  ',
+      }
+      const { errors } = await mutate({ mutation: updateUserMutation, variables })
+      expect(errors[0]).toHaveProperty(
+        'message',
+        'child "name" fails because ["name" length must be at least 3 characters long]',
+      )
     })
   })
+})
 
-  describe('DeleteUser', () => {
-    let deleteUserVariables
-    let asAuthor
-    const deleteUserMutation = gql`
-      mutation($id: ID!, $resource: [Deletable]) {
-        DeleteUser(id: $id, resource: $resource) {
+describe('DeleteUser', () => {
+  const deleteUserMutation = gql`
+    mutation($id: ID!, $resource: [Deletable]) {
+      DeleteUser(id: $id, resource: $resource) {
+        id
+        name
+        about
+        deleted
+        contributions {
           id
-          contributions {
-            id
-            deleted
-          }
+          content
+          contentExcerpt
+          deleted
           comments {
             id
+            content
+            contentExcerpt
             deleted
           }
         }
+        comments {
+          id
+          content
+          contentExcerpt
+          deleted
+        }
       }
-    `
+    }
+  `
+  beforeEach(async () => {
+    variables = { id: ' u343', resource: [] }
+
+    user = await factory.create('User', {
+      name: 'My name should be deleted',
+      about: 'along with my about',
+      id: 'u343',
+    })
+    await factory.create('User', {
+      email: 'friends-account@example.org',
+      password: '1234',
+      id: 'u565',
+    })
+  })
+
+  describe('unauthenticated', () => {
+    it('throws authorization error', async () => {
+      const { errors } = await mutate({ mutation: deleteUserMutation, variables })
+      expect(errors[0]).toHaveProperty('message', 'Not Authorised!')
+    })
+  })
+
+  describe('authenticated', () => {
     beforeEach(async () => {
-      await factory.create('User', {
-        email: 'test@example.org',
-        password: '1234',
-        id: 'u343',
-      })
-      await factory.create('User', {
-        email: 'friends-account@example.org',
-        password: '1234',
-        id: 'u565',
-      })
-      deleteUserVariables = { id: 'u343', resource: [] }
+      authenticatedUser = await user.toJson()
     })
 
-    describe('unauthenticated', () => {
-      it('throws authorization error', async () => {
-        client = new GraphQLClient(host)
-        await expect(client.request(deleteUserMutation, deleteUserVariables)).rejects.toThrow(
-          'Not Authorised',
-        )
+    describe("attempting to delete another user's account", () => {
+      beforeEach(() => {
+        variables = { ...variables, id: 'u565' }
```


this is a counter example where I needed to scroll up to see who `u565` is, maybe we could both be happy if instead of using variables at the highest-level, the id could be descriptive as you and others do in other places, like `not-my-account`